functions.py

```diff
@@ -0,0 +1,164 @@
+import smtplib
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from email.mime.application import MIMEApplication
+import os
+import pandas as pd
+import fitz
+from googleapiclient.discovery import build
+from google.oauth2.service_account import Credentials
+from PIL import Image
+from pdf2image import convert_from_bytes
+from io import BytesIO
+from dotenv import load_dotenv
+
+load_dotenv()
+# Function to generate a preview of the certificate
+
+def generate_preview(presentation_id, full_name):
+    creds = Credentials.from_service_account_info(
+        {
+            "type": "service_account",
+            "project_id": os.getenv("GOOGLE_PROJECT_ID"),
```

Comment:
Here we are fetching from environment variables twice (once in `generate_preview` and another time in `process_certificates`), the whole code is repeated twice, so instead of doing it inside function, we can create `creds`, `drive_service` and `slides_service` etc outside and remove repetition